src/solutions/optimization_solutions.jl

```diff
@@ -17,28 +17,43 @@ Representation of the solution to an nonlinear optimization defined by an Optimi
   callback (`sol.retcode === :Terminated`), or whether it exited due to an error. For more
   details, see the return code section of the DifferentialEquations.jl documentation.
 """
-struct OptimizationSolution{T, N, uType, P, A, Tf, O} <: AbstractOptimizationSolution{T, N}
+struct OptimizationSolution{T, N, uType, C, A, Tf, O} <: AbstractOptimizationSolution{T, N}
     u::uType # minimizer
-    prob::P # optimization problem
+    cache::C # optimization cache
     alg::A # algorithm
     minimum::Tf
     retcode::ReturnCode.T
     original::O # original output of the optimizer
 end
 
-function build_solution(prob::AbstractOptimizationProblem,
-                        alg, u, minimum;
-                        retcode = ReturnCode.Default,
-                        original = nothing,
-                        kwargs...)
+function Base.getproperty(sol::OptimizationSolution, sym::Symbol)
+    if sym == :prob
+        Base.depwarn("`sol.prob` is deprecated. Use `sol.cache` instead.", "sol.prob")
+        return getfield(sol, :cache)
```

Comment:
shouldn't this be `sol.cache.prob`?

Reply:
The cache does not hold the problem. It think it would be hard to keep in sync. But if you think it should, we can add it. 
Maybe @Vaibhavdixit02 has an additional insight.

Reply:
I think we have to ditch the `sol.prob.p` or `sol.cache.p` altogether and have to use a `get_p` interface.

For the `MOI` solvers there needs to be a `MOI.AbstractNLPEvaluator` which hold `p` but we need a cache that is a `SciMLBase.AbstractOptimizationCache`. This can only be solved via compostion, that the cache holds the evaluator. 


```julia
mutable struct MOIOptimizationNLPEvaluator{T, F <: OptimizationFunction, uType, P, LB, UB, I,
    JT <: DenseOrSparse{T}, HT <: DenseOrSparse{T},
    CHT <: DenseOrSparse{T}, S} <: MOI.AbstractNLPEvaluator
    f::F
    u0::uType
    p::P
    lb::LB
    ub::UB
    int::I
    J::JT
    H::HT
    cons_H::Vector{CHT}
    lcons::Vector{T}
    ucons::Vector{T}
    sense::S
end

mutable struct MOIOptimizationNLPCache{E, O, OS} <: SciMLBase.AbstractOptimizationCache
    evaluator::E
    opt::O
    opt_setup::OS
end
```